Lib/test/test_threading.py

```diff
@@ -1035,6 +1035,125 @@ def import_threading():
         self.assertEqual(out, b'')
         self.assertEqual(err, b'')
 
+    @cpython_only
+    def test_finalize_daemon_thread_hang(self):
+        # bpo-42969: tests that daemon threads hang during finalization
+        script = textwrap.dedent('''
+            import os
+            import sys
+            import threading
+            import time
+            import _testcapi
+
+            lock = threading.Lock()
+            lock.acquire()
+            def thread_func():
+                try:
+                    _testcapi.finalize_thread_hang(lock.acquire)
+                finally:
+                    # Control must not reach here.
+                    os._exit(2)
+
+            t = threading.Thread(target=thread_func)
+            t.daemon = True
+            t.start()
+            # Sleep to ensure daemon thread is blocked on `lock.acquire`
+            #
+            # Note: This test is designed so that in the unlikely case that
+            # `0.1` seconds is not sufficient time for the thread to become
+            # blocked on `lock.acquire`, the test will still pass, it just won't
+            # be properly testing the thread behavior during finalization.
+            time.sleep(0.1)
+
+            def run_during_finalization():
+                # Wake up daemon thread
+                lock.release()
+                # Sleep to give the daemon thread time to crash if it is going to.
+                time.sleep(0.1)
+                # If control reaches here, the test succeeded.
+                os._exit(0)
+
+            # Replace sys.stderr.flush as a way to run code during finalization
+            orig_flush = sys.stderr.flush
+            def do_flush(*args, **kwargs):
+                orig_flush(*args, **kwargs)
+                if not sys.is_finalizing:
+                    return
+                sys.stderr.flush = orig_flush
+                run_during_finalization()
+
+            sys.stderr.flush = do_flush
+
+            # If the follow exit code is retained, `run_during_finalization`
+            # did not run.
+            sys.exit(1)
+        ''')
+        assert_python_ok("-c", script)
+
+    @cpython_only
+    def test_finalize_block(self):
+        # bpo-42969: tests `PyThread_{Acquire,Release}FinalizeBlock`
+        script = textwrap.dedent('''
+            import os
+            import sys
+            import threading
+            import time
+            import _testcapi
+
+            sem = threading.Semaphore(0)
+            def thread_func():
+                # Should not return False
+                if not _testcapi.acquire_finalize_block():
+                    os._exit(5)
+                try:
+                    sem.release()
+                    # Sleep to allow Python interpreter to begin exiting.
+                    time.sleep(0.1)
```

Comment:
Sleep is a bad sync primitive: https://pythondev.readthedocs.io/unstable_tests.html#don-t-use-sleep-as-synchronization

Use an event or anything else.

Reply:
While I agree that the guideline is generally good, I think it should be amended with a caveat that for testing race conditions sometimes sleeping is the best option, and the test should be designed so that if the sleep interval is insufficient, the test passes (without properly exercising the race condition) rather than failing.

I have modified these tests to use `threading.Event` in addition to sleep.  I think just using threading.Event would make the tests less likely to exercise the race condition.